File: sdks/python/src/pricewatcha/api_error.py
"""Parse structured API v1 error responses."""

from __future__ import annotations

from typing import Any


def is_api_error_body(value: Any) -> bool:
    return (
        isinstance(value, dict)
        and isinstance(value.get("code"), str)
        and isinstance(value.get("message"), str)
        and (isinstance(value.get("http_status"), int) or value.get("http_status") is None)
        and isinstance(value.get("retry_recommended"), bool)
    )


def parse_api_error_payload(payload: Any) -> tuple[str, dict[str, Any] | None]:
    """Return (message, structured error body or None)."""
    if isinstance(payload, dict):
        err = payload.get("error")
        if is_api_error_body(err):
            return str(err["message"]), err
        detail = payload.get("detail")
        if isinstance(detail, str):
            return detail, None
        if detail is not None:
            return str(detail), None
    return "Request failed", None


def api_error_from_job_failure(job_error: Any) -> dict[str, Any] | None:
    if is_api_error_body(job_error):
        return job_error
    if isinstance(job_error, str) and job_error.strip():
        return {
            "code": "internal_error",
            "message": job_error,
            "http_status": 422,
            "retry_recommended": False,
            "retry_after_seconds": None,
        }
    return None

File: sdks/typescript/src/api-error.ts
/** Structured error body returned by Pricewatcha API v1. */

export interface ApiErrorBody {
  code: string;
  message: string;
  http_status: number | null;
  retry_recommended: boolean;
  retry_after_seconds?: number | null;
}

export function isApiErrorBody(value: unknown): value is ApiErrorBody {
  if (!value || typeof value !== "object") return false;
  const obj = value as Record<string, unknown>;
  return (
    typeof obj.code === "string" &&
    typeof obj.message === "string" &&
    (typeof obj.http_status === "number" || obj.http_status === null) &&
    typeof obj.retry_recommended === "boolean"
  );
}

/** Parse API JSON error payloads (structured `error` or legacy FastAPI `detail`). */
export function parseApiErrorPayload(payload: unknown): {
  message: string;
  errorBody?: ApiErrorBody;
} {
  if (payload && typeof payload === "object") {
    const record = payload as Record<string, unknown>;
    if (isApiErrorBody(record.error)) {
      return { message: record.error.message, errorBody: record.error };
    }
    if (typeof record.detail === "string") {
      return { message: record.detail };
    }
    if (record.detail != null) {
      return { message: String(record.detail) };
    }
  }
  return { message: "Request failed" };
}

export function apiErrorFromJobFailure(jobError: unknown): ApiErrorBody | undefined {
  if (isApiErrorBody(jobError)) {
    return jobError;
  }
  if (typeof jobError === "string" && jobError.trim()) {
    return {
      code: "internal_error",
      message: jobError,
      http_status: 422,
      retry_recommended: false,
      retry_after_seconds: null,
    };
  }
  return undefined;
}
